radicle-tracker/src/thread.rs

```diff
@@ -0,0 +1,348 @@
+use nonempty::NonEmpty;
+use rose_tree::{NodeIndex, RoseTree, ROOT};
+use std::{collections::HashMap, hash};
+
+/// Laws:
+/// Thread::new(comment).first() === comment
+/// thread.reply(comment).delete(comment) === thread
+/// Thread::new(comment).delete(comment) === None
+/// Thread::new(comment).edit(f, comment) === Thread::new(f(comment).unwrap())
+#[derive(Debug, Clone)]
+pub struct Thread<A> {
+    lut: HashMap<Path, NodeIndex>,
+    tree: RoseTree<A>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    MissingPath(Path),
+    MissingNode,
+}
+
+#[derive(Debug, PartialEq, Eq, hash::Hash, Clone)]
+pub struct Path(Vec<u32>);
+
+impl Path {
+    pub fn new(index: u32) -> Self {
+        Path(vec![index])
+    }
+
+    pub fn push(&mut self, index: u32) {
+        self.0.push(index);
+    }
+
+    pub fn pop(&mut self) -> Option<u32> {
+        self.0.pop()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub(crate) fn max_node(paths: &[Self]) -> u32 {
+        let mut max_n = 0;
+
+        for path in paths {
+            let last = path.clone().pop();
+            max_n = max_n.max(last.unwrap_or(max_n));
+        }
+
+        max_n
+    }
+
+    // Get the prefix keys of path that differ by 1
+    pub(crate) fn prefix_keys<'a, P>(&self, paths: P) -> Vec<Path>
+    where
+        P: Iterator<Item = &'a Path>,
+    {
+        let mut prefixes = vec![];
+        for path in paths {
+            let mut prefix = path.clone();
+            prefix.pop();
+            if *self == prefix {
+                prefixes.push(path.clone())
+            }
+        }
+
+        prefixes
+    }
+}
+
+impl From<Vec<u32>> for Path {
+    fn from(path: Vec<u32>) -> Self {
+        Path(path)
+    }
+}
+
+impl<A> Thread<A> {
+    /// Create a new `Thread` with `a` as the root of the `Thread`.
+    ///
+    /// The return value includes the [`Path`] to reach the root.
+    /// This should always be equal to `Path::new(0)`, and can
+    /// be used to [`Thread::view`] the value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use radicle_tracker::{Path, Thread};
+    ///
+    /// let (thread, root_path) = Thread::new(String::from("Discussing rose trees"));
+    ///
+    /// assert_eq!(thread.view(&root_path), Ok(&String::from("Discussing rose trees")));
+    /// assert_eq!(root_path, Path::new(0));
+    /// ```
+    pub fn new(a: A) -> (Self, Path)
+    where
+        A: Eq,
+    {
+        let (tree, root) = RoseTree::new(a);
+        let mut lut = HashMap::new();
+        let path = Path::new(0);
```

Comment:
Maybe more self-documenting to have a `ROOT` const under `Path`, so you can do `let path = Path::ROOT`. It just requires `Path::new` to be a `const fn`